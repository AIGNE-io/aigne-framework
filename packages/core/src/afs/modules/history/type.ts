import type { AFSEntry } from "../../type.js";

export interface HistoryStorageCreateEntry extends Omit<AFSEntry, "id"> {}

export interface HistoryStorageListOptions {
  filter?: {
    userId?: string;
    sessionId?: string;
  };
  limit?: number;
  orderBy?: [string, "asc" | "desc"][];
}

export abstract class HistoryStorage {
  abstract create(entry: HistoryStorageCreateEntry): Promise<AFSEntry>;

  abstract list(options?: HistoryStorageListOptions): Promise<{ list: AFSEntry[] }>;

  abstract read(path: string): Promise<AFSEntry | undefined>;
}
