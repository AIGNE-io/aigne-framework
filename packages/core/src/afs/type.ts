import type { Emitter } from "strict-event-emitter";
import type { Context, ContextEventMap } from "../aigne/context.js";
import type { AFSStorage } from "./storage/type.js";

export interface AFSListOptions {
  filter?: {
    userId?: string;
    sessionId?: string;
  };
  recursive?: boolean;
  maxDepth?: number;
  limit?: number;
  orderBy?: [string, "asc" | "desc"][];
}

export interface AFSSearchOptions {
  limit?: number;
}

export interface AFSWriteEntryPayload extends Omit<AFSEntry, "id" | "path"> {}

export interface AFSModule {
  readonly moduleId: string;

  readonly path: string;

  onMount?(root: AFSRoot): void;

  list?(path: string, options?: AFSListOptions): Promise<{ list: AFSEntry[] }>;

  read?(path: string): Promise<AFSEntry | undefined>;

  write?(path: string, content: AFSWriteEntryPayload): Promise<AFSEntry>;

  search?(path: string, query: string, options?: AFSSearchOptions): Promise<{ list: AFSEntry[] }>;
}

export type AFSRootEvents = ContextEventMap & {
  historyCreated: [{ context: Context; entry: AFSEntry }];
};

export interface AFSRoot extends Emitter<AFSRootEvents>, AFSModule {
  storage(module: AFSModule): AFSStorage;
}

export interface AFSEntry<T = any> {
  id: string;
  createdAt?: Date;
  updatedAt?: Date;
  path: string;
  userId?: string | null;
  sessionId?: string | null;
  summary?: string | null;
  metadata?: Record<string, any> | null;
  linkTo?: string | null;
  content?: T;
}
