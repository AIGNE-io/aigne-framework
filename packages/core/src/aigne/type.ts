import type { Agent } from "../agents/agent.ts";

export interface AIGNECLIAgents {
  chat?: Agent;

  agents?: AIGNECLIAgent[];
}

export interface AIGNECLIAgent {
  agent?: Agent;
  name?: string;
  alias?: string[];
  description?: string;
  commands?: AIGNECLIAgent[];
}
