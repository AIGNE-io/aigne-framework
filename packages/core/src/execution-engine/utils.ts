import type { FunctionAgentFn, Message } from "../agents/agent.js";
import type { Context, Runnable } from "./context.js";

export function sequential(..._agents: [Runnable, ...Runnable[]]): FunctionAgentFn {
  let agents = [..._agents];

  return async (input: Message, context?: Context) => {
    if (!context) throw new Error("Context is required for sequential agents");

    const output: Message = {};

    // Clone the agents to run, so that we can update the agents list during the loop
    const agentsToRun = [...agents];
    agents = [];

    for (const agent of agentsToRun) {
      const [o, transferToAgent] = await context.call(
        agent,
        { ...input, ...output },
        { returnActiveAgent: true },
      );
      Object.assign(output, o);
      agents.push(transferToAgent);
    }

    return output;
  };
}

export function parallel(..._agents: [Runnable, ...Runnable[]]): FunctionAgentFn {
  let agents = [..._agents];

  return async (input: Message, context?: Context) => {
    if (!context) throw new Error("Context is required for parallel agents");

    const result = await Promise.all(
      agents.map((agent) => context.call(agent, input, { returnActiveAgent: true })),
    );

    agents = result.map((i) => i[1]);
    const outputs = result.map((i) => i[0]);

    return Object.assign({}, ...outputs);
  };
}
