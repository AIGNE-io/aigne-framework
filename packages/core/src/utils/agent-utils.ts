import type { AgentHooks } from "../agents/agent.ts";
import type { AIGNECLIAgent, AIGNECLIAgents } from "../aigne/type.js";

const priorities: NonNullable<AgentHooks["priority"]>[] = ["high", "medium", "low"];

export function sortHooks(hooks: AgentHooks[]): AgentHooks[] {
  return hooks
    .slice(0)
    .sort(({ priority: a = "low" }, { priority: b = "low" }) =>
      a === b ? 0 : priorities.indexOf(a) - priorities.indexOf(b),
    );
}

export interface CLIAgent<T> {
  agent?: T;
  name?: string;
  alias?: string[];
  description?: string;
  commands?: CLIAgent<T>[];
}

export function mapCliAgent<A, O>(
  { agent, commands, ...input }: CLIAgent<A>,
  transform: (input: A) => O,
): CLIAgent<O> {
  return {
    ...input,
    agent: agent ? transform(agent) : undefined,
    commands: commands?.map((item) => mapCliAgent(item, transform)),
  };
}

export function findCliAgent(cli: AIGNECLIAgents, parent: string[], name: string) {
  let currentAgents: AIGNECLIAgent[] = cli.agents ?? [];
  for (const name of parent) {
    const found = currentAgents.find((i) => (i.name || i.agent?.name) === name);
    if (!found) throw new Error(`Agent ${name} not found in parent path ${parent.join(" -> ")}`);
    if (found.commands) currentAgents = found.commands;
    else currentAgents = [];
  }
  return currentAgents.find((i) => (i.name || i.agent?.name) === name)?.agent;
}
