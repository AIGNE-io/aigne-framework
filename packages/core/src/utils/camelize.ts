import camelize from "camelize-ts";

export function customCamelize<T extends Record<string, unknown>, K extends KeyofUnion<T> = never>(
  obj: T,
  { shallowKeys = [] }: { shallowKeys?: K[] } = {},
): CustomCamelize<T, K> {
  const shallow = Object.fromEntries(
    shallowKeys?.filter((key) => key in obj).map((key) => [key, obj[key as keyof T]]) ?? [],
  );
  const deep = Object.fromEntries(
    Object.entries(obj).filter(([key]) => !shallowKeys?.includes(key as K)),
  );

  return {
    ...camelize(shallow, true),
    ...camelize(deep),
  } as ReturnType<typeof customCamelize<T, K>>;
}

type KeyofUnion<U> = U extends Record<string, unknown> ? keyof U : never;

type CamelCase<S extends string> = S extends `${infer P1}_${infer P2}${infer P3}`
  ? `${P1}${Uppercase<P2>}${CamelCase<P3>}`
  : S;

const _unique: unique symbol = Symbol();
type _never = typeof _unique;

type ExtractTypeFromUnion<T, U> = Extract<T, U> extends never
  ? _never
  : Extract<T, U> extends Array<infer E>
    ? Array<E>
    : U;

type ExtractTypeWithConstFromUnion<T, U> = Exclude<T, Exclude<T, U>>;

export type CustomCamelize<
  T,
  ShallowKeys extends KeyofUnion<T> | undefined = undefined,
> = ExtractTypeFromUnion<T, never> extends never
  ? never
  : ExtractTypeFromUnion<T, Date> extends Date
    ? ExtractTypeWithConstFromUnion<T, Date> | CustomCamelize<Exclude<T, Date>>
    : ExtractTypeFromUnion<T, RegExp> extends RegExp
      ? ExtractTypeWithConstFromUnion<T, RegExp> | CustomCamelize<Exclude<T, RegExp>>
      : ExtractTypeFromUnion<T, Array<unknown>> extends Array<infer U>
        ? Array<CustomCamelize<U>> | CustomCamelize<Exclude<T, Array<U>>>
        : // biome-ignore lint/complexity/noBannedTypes: <explanation>
          ExtractTypeFromUnion<T, Function> extends Function
          ? // biome-ignore lint/complexity/noBannedTypes: <explanation>
              | ExtractTypeWithConstFromUnion<T, Function>
              // biome-ignore lint/complexity/noBannedTypes: <explanation>
              | CustomCamelize<Exclude<T, Function>>
          : ExtractTypeFromUnion<T, object> extends object
            ? {
                [K in keyof T as Uncapitalize<CamelCase<string & K>>]: K extends Exclude<
                  ShallowKeys,
                  undefined
                >
                  ? T[K]
                  : CustomCamelize<T[K]>;
              }
            : T;
