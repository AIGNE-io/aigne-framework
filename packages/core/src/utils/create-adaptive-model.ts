import { ClaudeChatModel, type ClaudeChatModelOptions } from "../models/claude-chat-model.js";
import { OpenAIChatModel, type OpenAIChatModelOptions } from "../models/openai-chat-model.js";
import { XAIChatModel, type XAIChatModelOptions } from "../models/xai-chat-model.js";

export type ModelProvider = "openai" | "anthropic" | "xai";

const { MODEL_PROVIDER = "openai", MODEL_NAME = "gpt-4o-mini" } = process.env;

export function createAdaptiveModel(
  provider?: ModelProvider | null,
  options?: OpenAIChatModelOptions | XAIChatModelOptions | ClaudeChatModelOptions,
) {
  const mergedProvider = provider ?? MODEL_PROVIDER;
  const mergedOptions = { model: MODEL_NAME, ...options };

  switch (mergedProvider) {
    case "openai":
      return new OpenAIChatModel(mergedOptions);
    case "anthropic":
      return new ClaudeChatModel(mergedOptions);
    case "xai":
      return new XAIChatModel(mergedOptions);
    default:
      throw new Error(`Unsupported model provider: ${mergedProvider}`);
  }
}
