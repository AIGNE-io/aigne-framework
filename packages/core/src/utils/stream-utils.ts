import type {
  AgentProcessAsyncGeneratorResult,
  AgentResponseChunk,
  AgentResponseStream,
  Message,
} from "../agents/agent.js";

export function objectToAgentResponseStream<T extends Message>(json: T): AgentResponseStream<T> {
  return new ReadableStream({
    start(controller) {
      controller.enqueue({ delta: { json } });
      controller.close();
    },
  });
}

function mergeResult<T extends Message>(output: T, chunk: AgentResponseChunk<T>) {
  if (chunk.delta.text) {
    for (const [key, text] of Object.entries(chunk.delta.text)) {
      const original = output[key] as string | undefined;
      const t = (original || "") + (text || "");
      if (t) Object.assign(output, { [key]: t });
    }
  }

  if (chunk.delta.json) {
    Object.assign(
      output,
      Object.fromEntries(Object.entries(chunk.delta.json).filter(([_, v]) => v !== undefined)),
    );
  }
}

export async function agentResponseStreamToObject<T extends Message>(
  stream: AgentResponseStream<T> | AgentProcessAsyncGeneratorResult<T>,
): Promise<T> {
  const json: T = {} as T;

  if (stream instanceof ReadableStream) {
    const reader = stream.getReader();
    for (;;) {
      const { value, done } = await reader.read();
      if (done) break;
      mergeResult(json, value);
    }
  } else {
    for (;;) {
      const chunk = await stream.next();
      if (chunk.value) {
        if (chunk.done) {
          Object.assign(json, chunk.value);
        } else {
          mergeResult(json, chunk.value);
        }
      }
      if (chunk.done) break;
    }
  }

  return json;
}

export function asyncGeneratorToReadableStream<T extends Message>(
  generator: AgentProcessAsyncGeneratorResult<T>,
): AgentResponseStream<T> {
  return new ReadableStream({
    async start(controller) {
      try {
        if (generator instanceof ReadableStream) {
          for await (const value of generator) {
            controller.enqueue(value);
          }
        } else {
          for (;;) {
            const chunk = await generator.next();
            if (chunk.value) {
              if (chunk.done) {
                controller.enqueue({ delta: { json: chunk.value } });
              } else {
                controller.enqueue(chunk.value);
              }
            }
            if (chunk.done) break;
          }
        }
      } catch (error) {
        controller.error(error);
      } finally {
        controller.close();
      }
    },
  });
}

export function onAgentResponseStreamEnd<T extends Message>(
  stream: AgentResponseStream<T>,
  callback: (result: T) => unknown,
) {
  return new ReadableStream({
    async start(controller) {
      try {
        const json: T = {} as T;

        const reader = stream.getReader();

        for (;;) {
          const { value, done } = await reader.read();
          if (done) break;

          controller.enqueue(value);

          mergeResult(json, value);
        }

        Object.assign(json, await callback(json));
        controller.enqueue({
          delta: { json },
        });
      } catch (error) {
        controller.error(error);
      } finally {
        controller.close();
      }
    },
  });
}

export function isAsyncGenerator<T extends AsyncGenerator>(
  value: AsyncGenerator | unknown,
): value is T {
  return typeof value === "object" && value !== null && Symbol.asyncIterator in value;
}
