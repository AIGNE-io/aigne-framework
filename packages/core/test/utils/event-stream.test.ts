import { expect, test } from "bun:test";
import type { AgentResponseChunk, Message } from "@aigne/core";
import {
  AgentResponseStreamParser,
  EventSourceParserStream,
} from "@aigne/core/utils/event-stream.js";
import { arrayToReadableStream, readableStreamToArray } from "@aigne/core/utils/stream-utils";

test("EventSourceParserStream should enqueue an error for invalid json", async () => {
  const stream = arrayToReadableStream([
    `data: {"delta": {"text": {"text": "hello"}}}\n\n`,
    `data: {"delta": {"text": {"text": " world"}}}\n\n`,
    "data: invalid json string\n\n",
  ]).pipeThrough(new EventSourceParserStream());

  expect(readableStreamToArray(stream)).resolves.toMatchSnapshot();
});

test("EventSourceParserStream should enqueue an error for error event", async () => {
  const stream = arrayToReadableStream([
    `data: {"delta": {"text": {"text": "hello"}}}\n\n`,
    `data: {"delta": {"text": {"text": " world"}}}\n\n`,
    `event: error\ndata: {"message": "test error"}\n\n`,
  ]).pipeThrough(new EventSourceParserStream());

  expect(readableStreamToArray(stream)).resolves.toMatchSnapshot();
});

test("AgentResponseStreamParser should raise error from upstream", async () => {
  const stream = arrayToReadableStream<AgentResponseChunk<Message>>([
    { delta: { text: { text: "hello" } } },
    { delta: { text: { text: " world" } } },
    new Error("should raise error"),
    { delta: { text: { text: " never" } } },
  ]).pipeThrough(new AgentResponseStreamParser());

  expect(readableStreamToArray(stream)).rejects.toThrow("should raise error");
});
