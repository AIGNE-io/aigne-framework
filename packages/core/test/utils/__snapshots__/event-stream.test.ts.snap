// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`EventStreamParser should enqueue an error for invalid json 1`] = `
[
  {
    "delta": {
      "text": {
        "text": "hello",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " world",
      },
    },
  },
  [Error: Parse response chunk json error: JSON Parse error: Unexpected identifier "invalid" invalid json string],
]
`;

exports[`EventStreamParser should enqueue an error for error event 1`] = `
[
  {
    "delta": {
      "text": {
        "text": "hello",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " world",
      },
    },
  },
  [Error: test error],
]
`;
